### PR TITLE
labelme_format: try harder to disambiguate from VOC

### LINKED ISSUE #658: Format detection fails to disambiguate labelme and voc on tfds voc2012
When I try to detect format of a tfds/voc2012, Datumaro detects both LabelMe and VOC. Probably, the problem is in XMLs with only image info inside, like this:

<details>

```
<annotation>
  <folder>2008</folder>
  <filename>2008_000001.jpg</filename>
  <source>
    <database>Unknown</database>
    <annotation>Unknown</annotation>
    <image>Unknown</image>
  </source>
  <segmented>0</segmented>
</annotation>
```

</details>

### PR DESCRIPTION
**Pull request by IRDonch**
### Summary
A dataset in the VOC format may contain XML files with no objects, which will be indistinguishable from LabelMe files. Instead of matching immediately in this case, keep checking files until either one of them can be definitively determined to be in the LabelMe format or not, or all files are exhausted.

Resolves #658.

### Checklist

- [x] I submit my changes into the `develop` branch
- [ ] I have added description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/datumaro/blob/develop/CHANGELOG.md)
- [ ] I have updated the [documentation](
  https://github.com/openvinotoolkit/datumaro/tree/develop/docs) accordingly
- [x] I have added tests to cover my changes
- [x] I have [linked related issues](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/openvinotoolkit/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2021 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: tests/assets/voc_dataset/voc_dataset2/ImageSets/Action/trainval.txt [hunk 1]
```diff
@@ -1,3 +1,4 @@
+@
```
Comment (zhiltsov-max) [inline tests/assets/voc_dataset/voc_dataset2/ImageSets/Action/trainval.txt:1]: What is the purpose in calling this file this way?

Comment (IRDonch) [reply tests/assets/voc_dataset/voc_dataset2/ImageSets/Action/trainval.txt:1]: It's so that it would come before A in the sorted list. I should probably add a comment about this.

File: tests/test_format_detection.py [hunk 2]
```diff
@@ -109,6 +109,34 @@ def detect(context):
         self.assertIn('*.txt', result.exception.failed_alternatives[0])
         self.assertIn('*.lst', result.exception.failed_alternatives[0])
 
+    @mark_requirement(Requirements.DATUM_GENERAL_REQ)
+    def test_require_files_success(self):
```
Comment (zhiltsov-max) [inline tests/test_format_detection.py:113]: I can't say it's a blocker, but consider splitting the detector tests into more specific ones. For example, this test currently covers at least 4 statements:
- sorting order of files
- exclusion of files
- confidence level on success 
- success itself

File: tests/assets/voc_dataset/voc_dataset2/ImageSets/Action/trainval.txt (hunk 1 above)
Comment (zhiltsov-max) [reply tests/assets/voc_dataset/voc_dataset2/ImageSets/Action/trainval.txt:1]: Can we have a dedicated test for this?

File: tests/test_format_detection.py (hunk 2 above)
Comment (IRDonch) [reply tests/test_format_detection.py:113]: I can only agree with regards to the 3rd bullet point - `test_empty_detector` already tests the default confidence level, so it's not necessary to check it in most other tests. I just added the asserts on the result value because I could. As for the rest:

* Success itself is impossible _not_ to test, since it's tested implicitly by virtue of having the call to `apply_format_detector` be unprotected by a `try` block.
* It's true that exclusion could be tested by a separate test. however, I didn't want to create a bunch of new tests, because the exclusion functionality is shared between `require_file` and `require_files` and so it's already tested by `require_file` tests. The reason I added an `exclude_fnames` parameter to the test at all is to make sure that the function forwards the value to the shared code.

Comment (IRDonch) [reply tests/test_format_detection.py:112]: I removed the redundant confidence level checks.

File: tests/assets/voc_dataset/voc_dataset2/ImageSets/Action/trainval.txt (hunk 1 above)
Comment (IRDonch) [reply tests/assets/voc_dataset/voc_dataset2/ImageSets/Action/trainval.txt:1]: I expanded the comment inside `@.xml`.

Comment (zhiltsov-max) [reply tests/assets/voc_dataset/voc_dataset2/ImageSets/Action/trainval.txt:1]: It's quite easy to overlook breaking changes, which will lead to regression in this question. There is no specific test for sorting, and if we remove sorting everything can work well depending on platform logic for returned values from file list fetching. VOC auto detection is a selling feature for some customers, so I think it's better to be protected with a bunch of tests around, stating the expected behavior and making sure we covered corner cases. I'll leave it up to you to decide on this, maybe it will be better to add such tests when we cover this case by some kind of logic described in my proposal.

Comment (IRDonch) [reply tests/assets/voc_dataset/voc_dataset2/ImageSets/Action/trainval.txt:1]: > There is no specific test for sorting, and if we remove sorting everything can work well depending on platform logic for returned values from file list fetching.

We unfortunately cannot decisively verify that sorting works, since we can't be sure that the filesystem on the testing machine doesn't give us the directory entries in sorted order to begin with. `ApplyFormatDetectorTest.test_require_files_success` is the best we can do WRT verifying this.

However, even if the sorting is accidentally broken and we don't catch it, that won't cause a regression with regards to #658. The sorting only exists to make the detector more deterministic (and thus make the tests more deterministic as well). For a regression to happen, the LabelMe detector would _also_ have to be broken (but in such a way that both `VocImportTest.test_can_detect_voc` and `LabelMeImporterTest.test_can_detect` still pass). So I think we're protected against regression as well as we can be.

Also, this argument doesn't really seem to support your original suggestion - a dedicated test would just have the same problem.

> VOC auto detection is a selling feature for some customers, so I think it's better to be protected with a bunch of tests around, stating the expected behavior and making sure we covered corner cases.

I don't necessarily disagree, but this isn't a general "improve VOC detection testing" PR - I'm just trying to fix a specific scenario (#658). And since I can piggyback on an existing test dataset, I don't really see why I shouldn't do that. I think the new test annotation file is reasonably documented, so there should be no confusion as to what it's for.

> The example is that we have only XML files with image metadata.

That would be a different scenario, so I think it's off-topic.
